tests/unit/einstein/test_ceph.py

```diff
@@ -0,0 +1,162 @@
+        While opening image which does not exist, this exception should be
+        raised
+        """
+        with self.assertRaises(file_system_exceptions.ImageNotFoundException):
+            bad_img = self.fs.get_image(TEST_NOT_EXIST_IMG_NAME)
+
+    def test_list_images(self):
+        """Test list mages"""
+        imgs = self.fs.list_images()
+        self.assertIn(CEPH_IMG, imgs)
+
+    def test_argument_out_of_range_exception(self):
+        """
+        Test InvalidArgument error in write, read, and discard functions
+        Error will be raised in write if part of the write would fall outside
+        the image
+        Error will be raised in read if part of the range specified is outside
+        the image Raises
+        """
+        self.assertRaises(file_system_exceptions.ArgumentsOutOfRangeException,
+                          self.fs.read, CEPH_IMG, RANGE_OUTSIDE_THE_IMAGE,
+                          len(TEST_DATA))
+        self.assertRaises(file_system_exceptions.ArgumentsOutOfRangeException,
+                          self.fs.write, CEPH_IMG, TEST_DATA,
+                          RANGE_OUTSIDE_THE_IMAGE)
+        self.assertRaises(file_system_exceptions.ArgumentsOutOfRangeException,
+                          self.fs.discard, CEPH_IMG, RANGE_OUTSIDE_THE_IMAGE,
+                          len(TEST_DATA))
+
+    def test_write(self):
+        """Test write image"""
+        original_data = self.fs.read(CEPH_IMG, OFFSET, len(TEST_DATA))
+        bytes_written = self.fs.write(CEPH_IMG, TEST_DATA, OFFSET)
+        data_read = self.fs.read(CEPH_IMG, OFFSET, bytes_written)
+        self.assertEqual(TEST_DATA, data_read)
+        self.fs.discard(CEPH_IMG, OFFSET, bytes_written)
+        data_after_read = self.fs.read(CEPH_IMG, OFFSET, bytes_written)
+        self.assertEqual(original_data, data_after_read)
+
+    def tearDown(self):
+        self.fs.remove(CEPH_IMG)
+
+
+class TestSnapshot(unittest.TestCase):
+    """Test snapshot operations"""
+    @trace
+    def setUp(self):
+        self.fs = ceph.RBD(_cfg.fs, _cfg.iscsi.password)
+        self.fs.create_image(CEPH_IMG, CEPH_IMG_SIZE)
+        self.fs.snap_image(CEPH_IMG, CEPH_SNAP_IMG)
+
+    def test_snap_image(self):
+        """Test create snapshot"""
+        snapList = self.fs.list_snapshots(CEPH_IMG)
+        self.assertIn(CEPH_SNAP_IMG, snapList)
+
+    def test_image_exist_exception(self):
+        """this exception will be raised if snap image already exists"""
+        self.assertRaises(file_system_exceptions.ImageExistsException,
+                          self.fs.snap_image, CEPH_IMG, CEPH_SNAP_IMG)
+
+    def snap_protect(self):
+        """protecting a snapshot"""
+        self.fs.snap_protect(CEPH_IMG, CEPH_SNAP_IMG)
+        isProtected = self.fs.snap_is_protected(CEPH_IMG, CEPH_SNAP_IMG)
+        self.assertTrue(isProtected, "Failed to protect a snapshot!")
+
+    def test_clone(self):
+        """Test clone image"""
+        self.snap_protect()
```

Comment:
This protect should be part of setUp. If setUps are not overlapping then make a separate class.